File: CSharp/QuickSort.cs
using System;

namespace QuickSort {
    class Program {
        static int partition(int[] A, int start, int end) {
            int pivot = A[(start + end) / 2];
            int i = start;
            int j = end;

            while (i <= j) {
                while (A[i] < pivot)
                    i++;
                while (A[j] > pivot)
                    j--;
                if (i <= j) {
                    int temp = A[i];
                    A[i] = A[j];
                    A[j] = temp;

                    i++;
                    j--;
                }
            }

            return i;
        }

        static void qSort(int[] A, int start, int end) {
            if (start < end) {
                int temp = partition(A, start, end);

                qSort(A, start, temp - 1);
                qSort(A, temp, end);
            }
        }

        static void print(int[] A) {
            for (int i = 0; i < A.Length; ++i)
                Console.Write("{0} ", A[i]);
            Console.WriteLine();
        }

        static void Main(string[] args) {
            Random rnd = new Random();
            const int n = 10;
            int[] A = new int[n];

            for (int i = 0; i < n; ++i)
                A[i] = rnd.Next(10);

            print(A);
            qSort(A, 0, A.Length - 1);
            print(A);
        }
    }
}

File: Pascal/QuickSort.pas
program QuickSort;

const
    n = 10-1;

var
    A: array[0..n] of integer;
    i: integer;

procedure print();
var
    i: integer;
begin
    for i := 0 to n do
        Write(A[i], ' ');
    Writeln();
end;

function partition(start: integer; _end: integer):integer; 
var
    pivot, i, j, temp: integer;
begin
    pivot := A[(start + _end) div 2];
    i := start;
    j := _end;
    
    while i <= j do
    begin
        while A[i] < pivot do
            inc(i);
        while A[j] > pivot do
            dec(j);
        if i <= j then
        begin
            temp := A[i];
            A[i] := A[j];
            A[j] := temp;
            
            inc(i);
            dec(j);
        end;      
    end;
    
    partition := i;
end;

procedure qSort(start: integer; _end: integer);
var
    temp: integer;
begin
    if start < _end then
    begin
        temp := partition(start, _end);
        
        qSort(start, temp - 1);
        qSort(temp, _end);
    end;

end;

begin
    for i := 0 to n do
        A[i] := random(10);
    
    print();
    qSort(0, n);
    print();
end.